File: models/Gpt-3.5-turbo/cpp/code/problems/additional_problems/Swap_Game.cpp

#include <iostream>
#include <queue>
#include <set>
using namespace std;

struct State {
    int grid[3][3];
    int moves;
};

int main() {
    int grid[3][3];

    // Read the input grid
    for (int i = 0; i < 3; i++) {
        for (int j = 0; j < 3; j++) {
            cin >> grid[i][j];
        }
    }

    // Define the desired grid
    int desiredGrid[3][3] = {{1, 2, 3}, {4, 5, 6}, {7, 8, 9}};

    // Initialize the queue and visited set
    queue<State> q;
    set<State> visited;

    // Create the initial state
    State initialState;
    memcpy(initialState.grid, grid, sizeof(grid)); // Copy the grid
    initialState.moves = 0;

    // Enqueue the initial state
    q.push(initialState);
    visited.insert(initialState);

    // Perform BFS
    while (!q.empty()) {
        State currState = q.front();
        q.pop();

        // Check if the grid is in the desired state
        if (memcmp(currState.grid, desiredGrid, sizeof(desiredGrid)) == 0) {
            cout << currState.moves << endl;
            return 0;
        }

        // Generate all possible next states
        int dx[4] = {0, 0, 1, -1};
        int dy[4] = {1, -1, 0, 0};

        for (int i = 0; i < 3; i++) {
            for (int j = 0; j < 3; j++) {
                for (int k = 0; k < 4; k++) {
                    int ni = i + dx[k];
                    int nj = j + dy[k];

                    if (ni >= 0 && ni < 3 && nj >= 0 && nj < 3) {
                        State nextState = currState;
                        swap(nextState.grid[i][j], nextState.grid[ni][nj]);
                        nextState.moves++;

                        // Check if the next state has been visited before
                        if (visited.find(nextState) == visited.end()) {
                            q.push(nextState);
                            visited.insert(nextState);
                        }
                    }
                }
            }
        }
    }

    // If no desired state is found
    cout << -1 << endl;

    return 0;
}

File: models/Gpt-3.5-turbo/py/code/additional_problems/Swap_Game.py

from collections import deque

def swap_tiles(state, i, j):
    # Function to swap two tiles in the grid
    new_state = list(state)
    new_state[i], new_state[j] = new_state[j], new_state[i]
    return ''.join(new_state)

def solve_puzzle(grid):
    # Convert the grid into a string
    state = ''.join(str(num) for row in grid for num in row)

    # Initialize the queue and visited set
    queue = deque([(state, 0)])
    visited = set([state])

    # Define the possible moves
    moves = [[1, 3], [0, 2, 4], [1, 5],
             [0, 4, 6], [1, 3, 5, 7], [2, 4, 8],
             [3, 7], [4, 6, 8], [5, 7]]

    while queue:
        curr_state, depth = queue.popleft()
        
        # Check if the current state matches the goal state
        if curr_state == "123456789":
            return depth
        
        # Find the index of the empty tile (denoted by 9)
        empty_tile_idx = curr_state.index("9")

        # Generate all possible moves from the current state
        for move in moves[empty_tile_idx]:
            new_state = swap_tiles(curr_state, empty_tile_idx, move)
            
            # Check if the new state has been visited before
            if new_state not in visited:
                # Enqueue the new state and update the visited set
                queue.append((new_state, depth + 1))
                visited.add(new_state)

    return -1  # Return -1 if the goal state is not reachable

# Read the input grid
grid = [list(map(int, input().split())) for _ in range(3)]

# Solve the puzzle and print the minimum number of moves
print(solve_puzzle(grid))
